File: erettsegi/emelt/2026_may/java/Varos.java
import module java.base;

public class Varos {

    public static void main(String[] args) throws Exception {
        var autoAzonosito = System.console().readLine("1. Feladat: Kérem adja meg az autó azonosítóját! ");
        var autoValtozasSorok = Files.readAllLines(Path.of(autoAzonosito + ".txt"));

        var sebessegValtozasok = new ArrayList<SebessegValtozas>();
        sebessegValtozasok.add(new SebessegValtozas(0, 0, 0));

        for(var sor : autoValtozasSorok) {
            var darabolt = sor.split("\t");

            sebessegValtozasok.add(new SebessegValtozas(Integer.parseInt(darabolt[0]), Integer.parseInt(darabolt[1]), Integer.parseInt(darabolt[2])));
        }

        System.out.println("2. Feladat: A " + sebessegValtozasok.get(1).kezdetiIdo + ". mp-ben indult el, a megfigyelés végén " + sebessegValtozasok.getLast().sebessegAVegen + " m/s sebességgel haladt");

        var atlepteE = false;
        for(var valtozas : sebessegValtozasok) {
            if(valtozas.sebessegAVegen > 14) {
                atlepteE = true;
                break;
            }
        }

        System.out.println("3. Feladat: Az autó " + (atlepteE ? "átlépte" : "nem lépte át") + " a sebességhatárt");

        var leghosszabbAlltValtozas1 = (SebessegValtozas) null;
        var leghosszabbAlltValtozas2 = (SebessegValtozas) null;

        for(var i = 0; i < sebessegValtozasok.size() - 1; ++i) {
            var valtozas1 = sebessegValtozasok.get(i);
            var valtozas2 = sebessegValtozasok.get(i + 1);

            if(valtozas1.sebessegAVegen == 0) {
                var elteltIdo = valtozas2.kezdetiIdo - valtozas1.zaroIdo;
                var leghosszabbElteltIdo = leghosszabbAlltValtozas1 == null ? -1 : leghosszabbAlltValtozas2.kezdetiIdo - leghosszabbAlltValtozas1.zaroIdo;

                if(elteltIdo > leghosszabbElteltIdo) {
                    leghosszabbAlltValtozas1 = valtozas1;
                    leghosszabbAlltValtozas2 = valtozas2;
                 }
            }
        }

        System.out.println("4. Feladat: Leghosszabbatt állt idő " + leghosszabbAlltValtozas1.zaroIdo + " és " + leghosszabbAlltValtozas2.kezdetiIdo + " mp között volt");

        var vizsgalandoIdopont = Integer.parseInt(System.console().readLine("5. Feladat: Mikor vizsgáljuk a sebességet? "));
        var vizsgalandoI = -1;

        for(var i = 0; i < sebessegValtozasok.size() - 1; ++i) {
            if(egyAdottValtozasonBeluliIdoE(vizsgalandoIdopont, sebessegValtozasok.get(i)) ||
               ketValtozasKozottiIdoE(vizsgalandoIdopont, sebessegValtozasok.get(i), sebessegValtozasok.get(i + 1)) ||
               egyAdottValtozasonBeluliIdoE(vizsgalandoIdopont, sebessegValtozasok.get(i + 1))) {

                vizsgalandoI = i;
                break;
             }
        }

        var sebessegAVizsgaltIdopontban = egyAdottValtozasonBeluliIdoE(vizsgalandoIdopont, sebessegValtozasok.get(vizsgalandoI)) ? sebessegAdottIdopontban(vizsgalandoIdopont, vizsgalandoI, sebessegValtozasok) :
                                          egyAdottValtozasonBeluliIdoE(vizsgalandoIdopont, sebessegValtozasok.get(vizsgalandoI + 1)) ? sebessegAdottIdopontban(vizsgalandoIdopont, vizsgalandoI + 1, sebessegValtozasok)
                                                                                                                                 : sebessegValtozasok.get(vizsgalandoI).sebessegAVegen;
        System.out.println("Az autó sebessége ekkor " + sebessegAVizsgaltIdopontban + "m/s volt");

        var osszesMegtettUt = 0.0F;
        for(var i = 0; i < sebessegValtozasok.size() - 1; ++i) {
            var valtozas1 = sebessegValtozasok.get(i);
            var valtozas2 = sebessegValtozasok.get(i + 1);

            osszesMegtettUt += valtozasonBelulMegtettUt(valtozas1, valtozas2) + valtozasKozottMegtettUt(valtozas1, valtozas2);
        }

        System.out.println("6. Feladat: A megtett út: " + osszesMegtettUt + "m");

        var fileba = new ArrayList<String>();
        for(var i = 0; i < sebessegValtozasok.size() - 1; ++i) {
            var valtozas1 = sebessegValtozasok.get(i);
            var valtozas2 = sebessegValtozasok.get(i + 1);

            fileba.add(valtozas2.kezdetiIdo + "\t" + valtozas1.sebessegAVegen);
            fileba.add(valtozas2.zaroIdo + "\t" + valtozas2.sebessegAVegen);
        }

        Files.write(Path.of("v" + autoAzonosito + ".txt"), fileba);

        // 8. Feladathoz nem vagyok hajlandó excelezni, sorry.
    }

    static boolean egyAdottValtozasonBeluliIdoE(int idopont, SebessegValtozas valtozas) {
        return idopont >= valtozas.kezdetiIdo && idopont <= valtozas.zaroIdo;
    }

    static boolean ketValtozasKozottiIdoE(int idopont, SebessegValtozas korabbiValtozas, SebessegValtozas kesobbiValtozas) {
        return idopont >= korabbiValtozas.zaroIdo && idopont <= kesobbiValtozas.kezdetiIdo;
    }

    static float sebessegAdottIdopontban(int t, int vizsgalandoI, ArrayList<SebessegValtozas> sebessegValtozasok) {
        var valtozas1 = sebessegValtozasok.get(vizsgalandoI - 1);
        var valtozas2 = sebessegValtozasok.get(vizsgalandoI);
        var v1 = (float) valtozas1.sebessegAVegen;
        var v2 = (float) valtozas2.sebessegAVegen;
        var t1 = (float) valtozas2.kezdetiIdo;
        var t2 = (float) valtozas2.zaroIdo;

        return v1 + ((v2 - v1) / (t2 - t1)) * (t - t1);
    }

    static float valtozasonBelulMegtettUt(SebessegValtozas valtozas1, SebessegValtozas valtozas2) {
        var v1 = (float) valtozas1.sebessegAVegen;
        var v2 = (float) valtozas2.sebessegAVegen;
        var t1 = (float) valtozas2.kezdetiIdo;
        var t2 = (float) valtozas2.zaroIdo;

        return (v1 + v2) / 2 * (t2 - t1);
    }

    static float valtozasKozottMegtettUt(SebessegValtozas valtozas1, SebessegValtozas valtozas2) {
        return valtozas1.sebessegAVegen * (valtozas2.kezdetiIdo - valtozas1.zaroIdo);
    }

    record SebessegValtozas(int kezdetiIdo, int zaroIdo, int sebessegAVegen) {}
}

File: erettsegi/emelt/2026_may/python/varos.py
from functools import cmp_to_key
from typing import NamedTuple

class SebessegValtozas(NamedTuple):
    kezdeti_ido: int
    zaro_ido: int
    sebesseg_a_vegen: int

def egy_adott_valtozason_beluli_ido_e(idopont: int, valtozas: SebessegValtozas):
    return idopont >= valtozas.kezdeti_ido and idopont <= valtozas.zaro_ido

def ket_valtozas_kozotti_ido_e(idopont: int, korabbiValtozas: SebessegValtozas, kesobbiValtozas: SebessegValtozas):
    return idopont >= korabbiValtozas.zaro_ido and idopont <= kesobbiValtozas.kezdeti_ido

def sebesseg_adott_idopontban(t: int, vizsgalandoI: int, sebessegValtozasok: list[SebessegValtozas]):
    valtozas1 = sebessegValtozasok[vizsgalandoI - 1]
    valtozas2 = sebessegValtozasok[vizsgalandoI]
    v1 = float(valtozas1.sebesseg_a_vegen)
    v2 = float(valtozas2.sebesseg_a_vegen)
    t1 = float(valtozas2.kezdeti_ido)
    t2 = float(valtozas2.zaro_ido)

    return v1 + ((v2 - v1) / (t2 - t1)) * (t - t1)

def valtozason_belul_megtett_ut(valtozas1: SebessegValtozas, valtozas2: SebessegValtozas):
    v1 = float(valtozas1.sebesseg_a_vegen)
    v2 = float(valtozas2.sebesseg_a_vegen)
    t1 = float(valtozas2.kezdeti_ido)
    t2 = float(valtozas2.zaro_ido)

    return (v1 + v2) / 2 * (t2 - t1)

def valtozas_kozott_megtett_ut(valtozas1: SebessegValtozas, valtozas2: SebessegValtozas):
    return valtozas1.sebesseg_a_vegen * (valtozas2.kezdeti_ido - valtozas1.zaro_ido)


auto_azonosito = input('1. Feladat: Kérem adja meg az autó azonosítóját! ')

with open(f'{auto_azonosito}.txt') as txt_input:
    valtozast_keszit = lambda k: SebessegValtozas(int(k[0]), int(k[1]), int(k[2]))

    sebesseg_valtozasok = [ valtozast_keszit(k.split('\t')) for k in txt_input.readlines() ]

sebesseg_valtozasok.insert(0, SebessegValtozas(0, 0, 0))

print(f'2. Feladat: A {sebesseg_valtozasok[1].kezdeti_ido}. mp-ben indult el, a megfigyelés végén {sebesseg_valtozasok[-1].sebesseg_a_vegen} m/s sebességgel haladt')

atlepte_e = any(k for k in sebesseg_valtozasok if k.sebesseg_a_vegen > 14)

print(f'3. Feladat: Az autó {"átlépte" if atlepte_e else "nem lépte át"} a sebességhatárt')

idot_osszehasonlit = lambda k1, k2: (k1[1].kezdeti_ido - k1[0].zaro_ido) - (k2[1].kezdeti_ido - k2[0].zaro_ido)
leghosszabb_allt_valtozas_paros = max((k for k in zip(sebesseg_valtozasok[: -1], sebesseg_valtozasok[1 :]) if k[0].sebesseg_a_vegen == 0), key = cmp_to_key(idot_osszehasonlit))

print(f'4. Feladat: Leghosszabbatt állt idő {leghosszabb_allt_valtozas_paros[0].zaro_ido} és {leghosszabb_allt_valtozas_paros[1].kezdeti_ido} mp között volt')

vizsgalando_idopont = int(input('5. Feladat: Mikor vizsgáljuk a sebességet? '))

vizsgalando_i = next(i for i in range(0, len(sebesseg_valtozasok) - 1) if \
        egy_adott_valtozason_beluli_ido_e(vizsgalando_idopont, sebesseg_valtozasok[i]) or
        ket_valtozas_kozotti_ido_e(vizsgalando_idopont, sebesseg_valtozasok[i], sebesseg_valtozasok[i + 1]) or
        egy_adott_valtozason_beluli_ido_e(vizsgalando_idopont, sebesseg_valtozasok[i + 1]))


sebesseg_a_vizsgalt_idopontban = sebesseg_adott_idopontban(vizsgalando_idopont, vizsgalando_i, sebesseg_valtozasok) if egy_adott_valtozason_beluli_ido_e(vizsgalando_idopont, sebesseg_valtozasok[vizsgalando_i]) else \
                                 sebesseg_adott_idopontban(vizsgalando_idopont, vizsgalando_i + 1, sebesseg_valtozasok) if egy_adott_valtozason_beluli_ido_e(vizsgalando_idopont, sebesseg_valtozasok[vizsgalando_i + 1]) else \
                                 sebesseg_valtozasok[vizsgalando_i].sebesseg_a_vegen

print(f'Az autó sebessége ekkor {sebesseg_a_vizsgalt_idopontban} m/s volt')

osszes_megtett_ut = sum(valtozason_belul_megtett_ut(k[0], k[1]) + valtozas_kozott_megtett_ut(k[0], k[1]) for k in zip(sebesseg_valtozasok[: -1], sebesseg_valtozasok[1 :]))

print(f'6. Feladat: A megtett út: {osszes_megtett_ut} m')

fileba = sum([[ f'{k[1].kezdeti_ido}\t{k[1].sebesseg_a_vegen}', f'{k[1].zaro_ido}\t{k[1].sebesseg_a_vegen}'] for k in zip(sebesseg_valtozasok[: -1], sebesseg_valtozasok[1 :])], [])

with open(f'v{auto_azonosito}.txt', 'w') as out_kiir:
    out_kiir.writelines(fileba)

# 8. Feladathoz nem vagyok hajlandó excelezni, sorry.
